proposals/function-pointers.md

```diff
@@ -11,122 +11,157 @@ potential implementation of the feature):
-- Custom attributes cannot be applied to a `funcptr` or any of its elements.
-- A `funcptr` parameter cannot be marked as `params`
-- A `funcptr` type has all of the restrictions of a normal pointer type.
+- Custom attributes cannot be applied to a `func*` or any of its elements.
+- A `func*` parameter cannot be marked as `params`
+- A `func*` type has all of the restrictions of a normal pointer type.
+
+### Function pointer syntax
+The full function pointer syntax is represented by the following grammar:
+
+``` 
+funcptr_type = 
+    'func' '*' [calling_convention] type method_arglist |
+    '(' funcptr_type ')' ;
+
+calling_convention = 
+    'managed' |
+    'unmanaged' |
+    'cdecl' |
+    'winapi' | 
+    'fastcall' | 
+    'stdcall' | 
+    'thiscall' ;
+```
+
+The `unmanaged` calling convention represents the default calling convention for native code on the current platform.
+
+When there is a nested function pointer, a function pointer which has or returns a function pointer, parens can be 
+opitionally used to disambiguate the signature. Though they are not required and the resulting types are equivalent.
+
+``` csharp
+delegate int Func1(string s);
+delegate Func1 Func2(Func1 f);
+
+// Function pointer equivalent without parens or calling convention
+func* int(string);
+func* func* int(string) int(func* int(string));
+
+// Function pointer equivalent without parens and with calling convention
+func* managed int(string);
+func* managed func* managed int(string) int(func* managed int(string));
+
+// Function pointer equivalent with parens and without calling convention
+func* int(string);
+func* (func* int(string)) int((func* int(string));
+
+// Function pointer equivalent of with parens and calling convention
+func* int(string)
+func* managed (func* managed int(string)) int((func* managed int(string));
+```
+
+When the calling convention is omitted from the syntax then `managed` will be used as the calling convention. That means
+all of the forms of `Func1` and `Func2` defined above are equivalent signatures.
+
+The calling convention cannot be omitted when the return type of the function pointer has the same name as a calling 
+convention. Inthat case the parser would process the return type as a calling convention instead of a type. To resolve
+this the developer must specify both the calling convention and the return type. 
+
+``` csharp
+class cdecl { }
+
+// Function pointer which has a cdecl calling convention, a cdecl return type and takes a single 
+// paramater of type cdecl;
+func* cdecl cdecl(cdecl);
+```
 
 ### Allow addresss-of to target methods
 
 Method groups will now be allowed as arguments to an address-of expression. The type of such an 
-expression will be an unnamed `funcptr` which has the equivalent signature of the target method and a managed 
+expression will be a `func*` which has the equivalent signature of the target method and a managed 
```

Comment:
What about taking the address of a method marked `extern` (`pinvokeimpl` in IL)? In this case, the method is known to be implemented in native code and may have an associated calling convention.

Reply:
See the section on `NativeCallbackAttribute`. That is fairly relevant to this scenario.

Reply:
I saw that section, after getting further down, as well. However, given that `DllImport` is not a custom attribute (at least from the perspective of the runtime/IL metadata) and that it encodes all the relevant information inline, it would be nice if C# just worked with the existing stuff.

For reference:
```csharp
[DllImport("a", BestFitMapping = false, CallingConvention = CallingConvention.Cdecl, CharSet = CharSet.Unicode, EntryPoint = "foo", ExactSpelling = true, PreserveSig = true, SetLastError =true, ThrowOnUnmappableChar =false)]
public static extern void Foo();
```

becomes:
```ilasm
method public hidebysig static pinvokeimpl("a" as "foo" nomangle unicode lasterr cdecl) 
    void Foo () cil managed preservesig 
{
}
```